File: legacy/packages/runtime/src/llm-trace.ts
import { mkdir, readdir, readFile, writeFile } from "fs/promises";
import { existsSync } from "fs";
import { join } from "path";
import type { ChatMessage } from "@moon/prompt";
import type { JsonSchema } from "@moon/schema-compiler";
import type { TokenUsage } from "./llm-types";

export interface LlmTraceEntry {
  id: string;
  timestamp: string;
  agent: string;
  model: string;
  stormRound?: number;
  delegateChain?: string[];
  messages: ChatMessage[];
  schema: JsonSchema;
  response?: unknown;
  error?: string;
  tokenUsage?: TokenUsage;
  durationMs: number;
}

export interface LlmTraceSession {
  runId: string;
  runDir: string;
  entries: LlmTraceEntry[];
}

export class LlmTraceWriter {
  private readonly entries: LlmTraceEntry[] = [];
  private seq = 0;

  constructor(
    public readonly runId: string,
    public readonly runDir: string,
  ) {}

  static async create(baseDir = join(process.cwd(), ".moon", "trace")): Promise<LlmTraceWriter> {
    const runId = new Date().toISOString().replace(/[:.]/g, "-");
    const runDir = join(baseDir, runId);
    await mkdir(runDir, { recursive: true });
    return new LlmTraceWriter(runId, runDir);
  }

  async record(entry: Omit<LlmTraceEntry, "id" | "timestamp">): Promise<void> {
    this.seq++;
    const id = String(this.seq).padStart(3, "0");
    const full: LlmTraceEntry = {
      ...entry,
      id,
      timestamp: new Date().toISOString(),
    };
    this.entries.push(full);
    const base = `${id}-${entry.agent}`;
    await writeFile(join(this.runDir, `${base}.json`), JSON.stringify(full, null, 2));
    await writeFile(join(this.runDir, `${base}-messages.txt`), formatMessagesText(full.messages));
    await writeFile(join(this.runDir, "manifest.json"), JSON.stringify({
      runId: this.runId,
      entries: this.entries.map((e) => ({ id: e.id, agent: e.agent, model: e.model })),
    }, null, 2));
  }

  snapshot(): LlmTraceSession {
    return { runId: this.runId, runDir: this.runDir, entries: [...this.entries] };
  }
}

function formatMessagesText(messages: ChatMessage[]): string {
  return messages.map((m) => `=== ${m.role.toUpperCase()} ===\n${m.content}\n`).join("\n");
}

export async function showLastTrace(baseDir?: string): Promise<string | null> {
  const root = baseDir ?? join(process.cwd(), ".moon", "trace");
  if (!existsSync(root)) return null;
  const runs = (await readdir(root, { withFileTypes: true }))
    .filter((d) => d.isDirectory())
    .map((d) => d.name)
    .sort();
  if (runs.length === 0) return null;
  return showTraceRun(runs[runs.length - 1]!, root);
}

export async function diffTraceRuns(
  runA: string,
  runB: string,
  baseDir?: string,
): Promise<string | null> {
  const root = baseDir ?? join(process.cwd(), ".moon", "trace");
  const dirA = join(root, runA);
  const dirB = join(root, runB);
  const manifestA = join(dirA, "manifest.json");
  const manifestB = join(dirB, "manifest.json");
  if (!existsSync(manifestA) || !existsSync(manifestB)) return null;

  const a = JSON.parse(await readFile(manifestA, "utf-8")) as { entries: Array<{ id: string; agent: string; model: string }> };
  const b = JSON.parse(await readFile(manifestB, "utf-8")) as { entries: Array<{ id: string; agent: string; model: string }> };

  const lines = [
    `Trace diff`,
    `  A: ${runA} (${a.entries.length} entries)`,
    `  B: ${runB} (${b.entries.length} entries)`,
    "",
  ];

  const agentsA = a.entries.map((e) => e.agent);
  const agentsB = b.entries.map((e) => e.agent);
  const onlyA = agentsA.filter((x) => !agentsB.includes(x));
  const onlyB = agentsB.filter((x) => !agentsA.includes(x));
  if (onlyA.length) lines.push(`Only in A: ${onlyA.join(", ")}`);
  if (onlyB.length) lines.push(`Only in B: ${onlyB.join(", ")}`);

  const shared = a.entries.filter((e) => agentsB.includes(e.agent));
  for (const entry of shared) {
    const other = b.entries.find((e) => e.agent === entry.agent);
    if (!other) continue;
    const msgA = join(dirA, `${entry.id}-${entry.agent}-messages.txt`);
    const msgB = join(dirB, `${other.id}-${other.agent}-messages.txt`);
    if (!existsSync(msgA) || !existsSync(msgB)) continue;
    const textA = await readFile(msgA, "utf-8");
    const textB = await readFile(msgB, "utf-8");
    if (textA !== textB) {
      lines.push("", `--- ${entry.agent} (messages differ)`);
      lines.push(`A bytes: ${textA.length}, B bytes: ${textB.length}`);
    }
  }

  if (lines.length <= 4 && onlyA.length === 0 && onlyB.length === 0) {
    lines.push("No differences in agent sets or shared message bodies.");
  }

  return lines.join("\n");
}

export async function showTraceRun(runId: string, baseDir?: string): Promise<string | null> {
  const runDir = join(baseDir ?? join(process.cwd(), ".moon", "trace"), runId);
  const manifestPath = join(runDir, "manifest.json");
  if (!existsSync(manifestPath)) return null;
  const manifest = JSON.parse(await readFile(manifestPath, "utf-8")) as { entries: Array<{ id: string; agent: string }> };
  const parts = [`Trace run: ${runId}`, `Directory: ${runDir}`, ""];
  for (const e of manifest.entries) {
    const msgPath = join(runDir, `${e.id}-${e.agent}-messages.txt`);
    if (existsSync(msgPath)) {
      parts.push(await readFile(msgPath, "utf-8"));
    }
  }
  return parts.join("\n");
}

File: Sources/MoonRuntime/LlmTrace.swift
import Foundation

public struct LlmTraceEntry: Codable, Sendable, Equatable {
    public var id: String
    public var timestamp: String
    public var agent: String
    public var model: String
    public var stormRound: Int?
    public var delegateChain: [String]?
    public var messages: [LlmChatMessage]
    public var schemaType: String?
    public var response: String?
    public var error: String?
    public var durationMs: Int
}

public struct LlmTraceSession: Sendable, Equatable {
    public var runId: String
    public var runDir: String
    public var entries: [LlmTraceEntry]
}

public final class LlmTraceWriter: @unchecked Sendable {
    public let runId: String
    public let runDir: String
    private var entries: [LlmTraceEntry] = []
    private var seq = 0
    private let lock = NSLock()

    public init(runId: String, runDir: String) {
        self.runId = runId
        self.runDir = runDir
    }

    public static func create(baseDir: String? = nil) async throws -> LlmTraceWriter {
        let root = baseDir ?? (FileManager.default.currentDirectoryPath as NSString)
            .appendingPathComponent(".moon/trace")
        let runId = ISO8601DateFormatter().string(from: Date())
            .replacingOccurrences(of: ":", with: "-")
            .replacingOccurrences(of: ".", with: "-")
        let runDir = (root as NSString).appendingPathComponent(runId)
        try FileManager.default.createDirectory(atPath: runDir, withIntermediateDirectories: true)
        return LlmTraceWriter(runId: runId, runDir: runDir)
    }

    public func record(
        agent: String,
        model: String,
        messages: [LlmChatMessage],
        schema: JsonSchema,
        response: RuntimeValue? = nil,
        error: String? = nil,
        stormRound: Int? = nil,
        delegateChain: [String]? = nil,
        durationMs: Int
    ) async throws {
        let entry: LlmTraceEntry = lock.withLock {
            seq += 1
            let id = String(format: "%03d", seq)
            let full = LlmTraceEntry(
                id: id,
                timestamp: ISO8601DateFormatter().string(from: Date()),
                agent: agent,
                model: model,
                stormRound: stormRound,
                delegateChain: delegateChain,
                messages: messages,
                schemaType: schema.typeName,
                response: response.map(runtimeValueDescription),
                error: error,
                durationMs: durationMs
            )
            entries.append(full)
            return full
        }

        let base = "\(entry.id)-\(entry.agent)"
        let jsonData = try JSONEncoder().encode(entry)
        try jsonData.write(to: URL(fileURLWithPath: (runDir as NSString).appendingPathComponent("\(base).json")))
        let messagesText = entry.messages.map { "=== \($0.role.uppercased()) ===\n\($0.content)\n" }.joined(separator: "\n")
        try messagesText.write(
            toFile: (runDir as NSString).appendingPathComponent("\(base)-messages.txt"),
            atomically: true,
            encoding: .utf8
        )

        let manifest: [String: Any] = [
            "runId": runId,
            "entries": entries.map { ["id": $0.id, "agent": $0.agent, "model": $0.model] },
        ]
        let manifestData = try JSONSerialization.data(withJSONObject: manifest, options: [.prettyPrinted])
        try manifestData.write(to: URL(fileURLWithPath: (runDir as NSString).appendingPathComponent("manifest.json")))
    }

    public func snapshot() -> LlmTraceSession {
        lock.withLock {
            LlmTraceSession(runId: runId, runDir: runDir, entries: entries)
        }
    }
}

public func showLastTrace(baseDir: String? = nil) throws -> String? {
    let root = baseDir ?? (FileManager.default.currentDirectoryPath as NSString)
        .appendingPathComponent(".moon/trace")
    guard FileManager.default.fileExists(atPath: root) else { return nil }
    let runs = try FileManager.default.contentsOfDirectory(atPath: root).sorted()
    guard let last = runs.last else { return nil }
    return try showTraceRun(last, baseDir: root)
}

public func showTraceRun(_ runId: String, baseDir: String? = nil) throws -> String? {
    let root = baseDir ?? (FileManager.default.currentDirectoryPath as NSString)
        .appendingPathComponent(".moon/trace")
    let runDir = (root as NSString).appendingPathComponent(runId)
    let manifestPath = (runDir as NSString).appendingPathComponent("manifest.json")
    guard FileManager.default.fileExists(atPath: manifestPath) else { return nil }
    let data = try Data(contentsOf: URL(fileURLWithPath: manifestPath))
    let manifest = try JSONSerialization.jsonObject(with: data) as? [String: Any]
    let entries = manifest?["entries"] as? [[String: Any]] ?? []
    var lines = ["Trace run: \(runId)", "Entries: \(entries.count)", ""]
    for entry in entries {
        let id = entry["id"] as? String ?? "?"
        let agent = entry["agent"] as? String ?? "?"
        let model = entry["model"] as? String ?? "?"
        lines.append("  \(id)  \(agent)  (\(model))")
    }
    return lines.joined(separator: "\n")
}

public func diffTraceRuns(_ runA: String, _ runB: String, baseDir: String? = nil) throws -> String? {
    let root = baseDir ?? (FileManager.default.currentDirectoryPath as NSString)
        .appendingPathComponent(".moon/trace")
    let manifestA = try Data(contentsOf: URL(fileURLWithPath: (root as NSString).appendingPathComponent("\(runA)/manifest.json")))
    let manifestB = try Data(contentsOf: URL(fileURLWithPath: (root as NSString).appendingPathComponent("\(runB)/manifest.json")))
    let a = try JSONSerialization.jsonObject(with: manifestA) as? [String: Any]
    let b = try JSONSerialization.jsonObject(with: manifestB) as? [String: Any]
    let entriesA = a?["entries"] as? [[String: Any]] ?? []
    let entriesB = b?["entries"] as? [[String: Any]] ?? []
    let agentsA = entriesA.compactMap { $0["agent"] as? String }
    let agentsB = entriesB.compactMap { $0["agent"] as? String }
    var lines = [
        "Trace diff",
        "  A: \(runA) (\(entriesA.count) entries)",
        "  B: \(runB) (\(entriesB.count) entries)",
        "",
    ]
    let onlyA = agentsA.filter { !agentsB.contains($0) }
    let onlyB = agentsB.filter { !agentsA.contains($0) }
    if !onlyA.isEmpty { lines.append("Only in A: \(onlyA.joined(separator: ", "))") }
    if !onlyB.isEmpty { lines.append("Only in B: \(onlyB.joined(separator: ", "))") }
    return lines.joined(separator: "\n")
}

private extension NSLock {
    func withLock<T>(_ body: () -> T) -> T {
        lock()
        defer { unlock() }
        return body()
    }
}